evm-ds/src/precompiles/scilla_read.rs

```diff
@@ -0,0 +1,252 @@
+                // Malformatted json
+                return Err(PrecompileFailure::Error {
+                    exit_status: ExitError::Other(Cow::Borrowed(
+                        "Unable to find field with given name in fields list",
+                    )),
+                });
+            }
+            field_types.push(substitute_scilla_type_with_sol(chunks[idx + 1])?);
+            idx += 2;
+        } else {
+            field_types.push(substitute_scilla_type_with_sol(chunk)?);
+            idx += 1;
+        }
+    }
+
+    let Ok(decoded_array) = decode(&field_types, input) else {
+        return Err(PrecompileFailure::Error {
+            exit_status: ExitError::Other(Cow::Borrowed("Unable to decode input by specified field definition")),
+        });
+    };
+
+    let mut indices = vec![];
+    for value in decoded_array.iter().skip(2) {
+        match value {
+            Token::Uint(solidity_uint) => {
+                indices.push(format!("{}", solidity_uint));
+            }
+            Token::Address(solidity_addr) => {
+                indices.push(format!("0x{}", hex::encode(solidity_addr)));
+            }
+            _ => {
+                indices.push(value.to_string());
+            }
+        }
+    }
+    Ok(ScillaField {
+        indices,
+        ret_type: chunks.last().unwrap().to_string(),
+    })
+}
+
+fn extract_substate_from_json(
+    value: &Value,
+    vname: &str,
+    def: &ScillaField,
+) -> Result<Option<Token>, PrecompileFailure> {
+    if !value.is_object() || value.get(vname).is_none() {
+        return Ok(None);
+    }
+    if def.indices.is_empty() {
+        return encode_result_type(&value[vname], def);
+    }
+    let mut value = value[vname].as_object().unwrap();
+
+    for index in def.indices.clone().iter().take(def.indices.len() - 1) {
+        if !value[index].is_object() {
+            return Err(PrecompileFailure::Error {
+                exit_status: ExitError::Other(Cow::Borrowed(
+                    "Scilla field definition doesn't match with returned value",
+                )),
+            });
+        }
+        value = value[index].as_object().unwrap();
+    }
+
+    encode_result_type(&value[def.indices.last().unwrap()], def)
+}
+
+fn encode_result_type(
+    value: &Value,
+    def: &ScillaField,
+) -> Result<Option<Token>, PrecompileFailure> {
+    if def.ret_type.starts_with("Uint") {
+        if value.is_null() || !value.is_string() {
+            return Ok(Some(Token::Uint(Uint::from(0))));
+        }
+        let value = value.as_str().unwrap();
+        return Ok(Some(Token::Uint(
+            Uint::from_dec_str(value).unwrap_or_default(),
+        )));
+    } else if def.ret_type.starts_with("Int") {
+        if value.is_null() || !value.is_string() {
+            return Ok(Some(Token::Int(Uint::from(0))));
+        }
+        let value = value.as_str().unwrap();
+        return Ok(Some(Token::Int(
+            Uint::from_dec_str(value).unwrap_or_default(),
```

Comment:
Does this not result in -ve numbers either being wrapped or ignored?

Reply:
Well, it seems that underneath there's typedef Int = ethereum_types::Uint256 anyway.